apps/yapms/src/lib/components/modals/toolsmodal/tools/tablecomponents/CandidateActions.svelte

```diff
@@ -0,0 +1,135 @@
+<script lang="ts">
+	import type { Region } from '$lib/types/Region';
+
+	import { get } from 'svelte/store';
+	import { TossupCandidateStore } from '$lib/stores/Candidates';
+	import { CandidatesStore } from '$lib/stores/Candidates';
+	import { RegionsStore } from '$lib/stores/regions/Regions';
+	import { makePatternCSS } from '$lib/utils/patterns';
+	import { splitRegion } from '$lib/stores/regions/regionActions';
+	import Sliders from '$lib/icons/Sliders.svelte';
+
+	export let region: Region;
+
+	$: maxValue = region.candidates.reduce(
+		(prev, current) => (prev > current.count ? prev : current.count),
+		region.candidates[0].count
+	);
+
+	$: winners = region.disabled
+		? [
+				{
+					candidate: $TossupCandidateStore,
+					count: 0,
+					margin: 0
+				}
+			]
+		: region.candidates.filter((candidate) => candidate.count === maxValue);
+
+	function getWinnerId(region: Region) {
+		//if region is split, then select value is "split"
+		if (checkIfSplit(region)) {
+			return 'split';
+		}
+
+		return winners[0].candidate.id;
+	}
+
+	function checkIfSplit(region: Region) {
+		for (const candidate of region.candidates) {
+			if (candidate.count == region.value) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	function getWinningColor(region: Region) {
+		// if the region fill is a pattern, get the corresponding CSS gradient.
+		if (region.nodes.region.style.fill.substring(0, 3) === 'url') {
+			return makePatternCSS(winners);
+		}
+		return region.nodes.region.style.fill;
+	}
+
+	function changeWinner(event: Event & { currentTarget: EventTarget & HTMLSelectElement }) {
+		const newWinnerId = event.currentTarget.value;
+
+		const candidates = get(CandidatesStore);
+		let newWinner = candidates.find((candidate) => candidate.id === newWinnerId);
+
+		const tossupCandidate = get(TossupCandidateStore);
+		if (newWinnerId == tossupCandidate.id) {
+			newWinner = tossupCandidate;
+		}
+
+		if (newWinner) {
+			const newCandidate = {
+				candidate: newWinner,
+				count: region.value,
+				margin: 0
+			};
+			region.candidates = [newCandidate];
+			$RegionsStore = $RegionsStore;
+		}
+	}
+
+	function cycleWinnerMargin(event: MouseEvent) {
+		if (winners.length != 1) {
+			return;
+		}
+
+		const candidateToBeEdited = winners[0];
+
+		const currentCandidateMargin = winners[0].margin;
+
+		if (!event.ctrlKey) {
+			candidateToBeEdited.margin =
+				currentCandidateMargin === candidateToBeEdited.candidate.margins.length - 1
+					? 0
+					: candidateToBeEdited.margin + 1;
+		} else {
+			candidateToBeEdited.margin =
+				currentCandidateMargin === 0
+					? candidateToBeEdited.candidate.margins.length - 1
+					: candidateToBeEdited.margin - 1;
+		}
+
+		region.candidates = [candidateToBeEdited];
+		$RegionsStore = $RegionsStore;
+	}
```

Comment:
I think this should be moved to /lib/stores/regions/regionActions.ts just to keep all the region manipulation logic in one spot. At least as much as possible.

I just copy & pasted the fillRegion function and added a variable for candidateID as a string. I think that can then be used in spot of your changeWinner and cylceWinnerMargin functions.

I think long term, the fillRegion function should be changed to take in a config object.

![image](https://github.com/user-attachments/assets/0bc28d23-9671-45c0-9749-1041fb2b070d)

Reply:
I was able to put the function here.

![image](https://github.com/user-attachments/assets/7c7a4b02-149e-4b73-8d74-05dd6afa78d9)

Reply:
Right, this is good. It certainly felt as if we were lacking proper standardized ways to manipulate regions as I was doing this. I will move the logic for updateRegionValue as well.

Reply:
Did find this is kinda slow due to the region search. Note to self for if I come back to this: function that just consumes a region object